runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+     * all range queries for the same stream are coalesced into a single request.
+     * For example, the following queries will all be merged into one query (i.e., (-1, 100])
+     *  Query1  (-1, 100)
+     *  Query2  (2, 5)
+     *  Query3  (3, 10)
+     */
+    private List<StreamAddressRange> coalesce(List<StreamAddressRange> ranges) {
+        Map<UUID, Interval> adjusted = new HashMap<>();
+
+        for (StreamAddressRange range : ranges) {
+            Interval interval = adjusted.getOrDefault(range.getStreamID(), new Interval(Long.MAX_VALUE, Long.MIN_VALUE));
+            Interval expandedInterval = merge(interval, range);
+            Preconditions.checkState(expandedInterval.x >= Address.NEVER_READ);
+            Preconditions.checkState(expandedInterval.x < expandedInterval.y);
+
+            adjusted.put(range.getStreamID(), expandedInterval);
+        }
+
+        List<StreamAddressRange> adjustedList = new ArrayList<>();
+        for (Map.Entry<UUID, Interval> entry : adjusted.entrySet()) {
+            adjustedList.add(new StreamAddressRange(entry.getKey(), entry.getValue().y, entry.getValue().x));
+        }
+        return adjustedList;
+    }
+
+    private void dispatchSyncTasks(List<StreamingTask> tasks) {
+        for (StreamingTask task : tasks) {
+            if (task.getStream().hasNext()) {
+                // this will fail for tasks that are already syncing
+                // for tasks to be scheduled to sync, they must be in the scheduling state
+                task.move(StreamStatus.SCHEDULING, StreamStatus.SYNCING);
+                workers.execute(task);
+            } else {
+                // no new deltas to sync
+                // This stream has no entries to sync after the poll, set it to runnable so that it can
+                // be polled next scheduling cycle
+                task.move(StreamStatus.SCHEDULING, StreamStatus.RUNNABLE);
+            }
+        }
+    }
+
+    private void handleFailures(List<StreamingTask> failedTasks) {
+        // Need to remove the tasks from allTasks before propagating the error just incase they try to
+        // re-register the task?
+        failedTasks.forEach(t -> allTasks.remove(t.getListener()));
+
+        for (StreamingTask task : failedTasks) {
+            workers.execute(task::propagateError);
+        }
+    }
+
+    public void schedule() throws InterruptedException {
+        // Block thread until there's work to do
+        waitForTasks();
+
+        final long cycleStartTimeNs = System.nanoTime();
+
+        final List<StreamingTask> runnableTasks = getTasks(StreamStatus.RUNNABLE);
+        final List<StreamingTask> syncingTasks = getTasks(StreamStatus.SYNCING,
+                t -> t.getStream().availableSpace() >= pollThreshold);
+        final List<StreamingTask> failedTasks = getTasks(StreamStatus.ERROR);
+
+        runnableTasks.forEach(t -> t.move(StreamStatus.RUNNABLE, StreamStatus.SCHEDULING));
+        poll(runnableTasks);
```

Comment:
Perhaps it's me, but I had a hard time understanding these 4 lines. Since you understand the code more if you think of some way that this is more intuitive I think it would help. Also, one question what is the value of the status 'scheduling'? is it equivalent to 'waiting to sync'?

Reply:
No it i means its waiting to be polled. 

This state can be eliminated from a single threaded scheduler, but the reason I introduced it is because it will make a multi-threaded scheduler implementation straight forward, if it is needed.